### ISSUE TITLE
Incorrect `index-out-of-bounds` error

### ISSUE THREAD

**Issue by alechouse97** [astral-sh/ty#2395]
### Summary

Ty is currently raising an `index-out-of-bounds` error when facing a block of code, for example in this `tmp.py`:

```python
def func(x: tuple[float, float] | tuple[float, float, float]) -> None:
    if len(x) == 2:
        y = x[2]
    elif len(x) == 3:
        y = x[3]
```

It correctly catches that `x[2]` is out of bounds in the first branch, but also states `x[2]` is out of bounds in the second branch:

```
error[index-out-of-bounds]: Index 2 is out of bounds for tuple `tuple[int | float, int | float]` with length 2
 --> tmp.py:3:13
  |
1 | def func(x: tuple[float, float] | tuple[float, float, float]):
2 |     if len(x) == 2:
3 |         y = x[2]
  |             ^
4 |     elif len(x) == 3:
5 |         y = x[2]
  |
info: rule `index-out-of-bounds` is enabled by default

error[index-out-of-bounds]: Index 2 is out of bounds for tuple `tuple[int | float, int | float]` with length 2
 --> tmp.py:5:13
  |
3 |         y = x[2]
4 |     elif len(x) == 3:
5 |         y = x[2]
  |             ^
  |
info: rule `index-out-of-bounds` is enabled by default
```

Expected behavior is that ty narrows the type to `tuple[float, float, float]` based on the current branch that required

This can reproduced when running `ty check tmp.py` on the above.

As an aside, it also looks to be changing the type to `tuple[int | float, int | float]`, but I've not been using ty long enough to know if that is expected.

### Version

ty 0.0.10

**Comment by AlexWaygood** [comment]
thanks!

**Comment by carljm** [comment]
> As an aside, it also looks to be changing the type to `tuple[int | float, int | float]`, but I've not been using ty long enough to know if that is expected.

See https://docs.astral.sh/ty/reference/typing-faq/#why-does-ty-show-int-float-when-i-annotate-something-as-float
